boofuzz/socket_connection.py

```diff
@@ -160,8 +160,18 @@ def open(self):
 
             if self.proto == "tcp" or self.proto == "ssl":
                 self._serverSock = self._sock
-                self._serverSock.listen(1)
-                self._sock, addr = self._serverSock.accept()
+                try:
+                    self._serverSock.listen(1)
+                    self._sock, addr = self._serverSock.accept()
+                except socket.error as e:
+                    # When connection timeout expires, tear down the server socket so we can re-open it again after
+                    # restarting the target.
+                    self._serverSock.shutdown(socket.SHUT_RDWR)
```

Comment:
Why are we using `shutdown()` here but nowhere else?
https://github.com/jtpereyda/boofuzz/blob/e3c79dc0257f10ee4bfaa7273b87d702d3a9d030/boofuzz/socket_connection.py#L116-L125
Should we use it everywhere or nowhere?

According to https://docs.python.org/3/library/socket.html#socket.socket.close using `shutdown()` would close the connection immediately, while `close()` only decrements the handle count and closes the socket when no handles are left. See https://stackoverflow.com/questions/409783/socket-shutdown-vs-socket-close

This might be related to https://github.com/jtpereyda/boofuzz/issues/296

Reply:
I can't comment on where else shutdown() might be needed, but I added it here specifically to get past errors for the port already being in use. I don't remember the specific error raised but I think it might have been the "address already in use" message from #296 which you mention. In general, if you're done with a connection for one test, and are closing it or letting it get garbage collected, but the next test trys to establish a connection on the same port, it probably is a good idea to call shutdown().

Reply:
Just checking in case I misunderstood something: is there something you wanted changed in this pull request before it's ready?

Reply:
It's fine by me, I'm just unsure whether it would be a benefit to add the socket shutdown to the default close function. If we want to make sure the connection is closed immediately, the shutdown method might help.

> **Note:** close() releases the resource associated with a connection but does not necessarily close the connection immediately. If you want to close the connection in a timely fashion, call shutdown() before close().

From https://docs.python.org/3/library/socket.html#socket.socket.close

Either way it's up to @jtpereyda to merge this.

Reply:
I didn't see anything to the effect that shutdown is safer in my usual sockets reference: http://beej.us/guide/bgnet/html/#close-and-shutdownget-outta-my-face

And I don't think anyone's encountered this error before with boofuzz (server fuzzing is a newer use case though). @ColBFritz I wonder if your error might have been related to some of the recent sockets related fixes. Would you be willing to switch it back to close (after merging in updates from master)? If the same errors happen again, please share them here so we can check it out. Thanks!

Reply:
I did some more research regarding the shutdown function and from what I read, it is considered best practice to use it. Though, it is almost never really required.
See https://docs.python.org/2/howto/sockets.html#disconnecting

As I understood it, `shutdown()` immediately sends a FIN but it does not free/close the socket.
Calling `close()` without a previous shutdown might not send the FIN right away but instead will try to finish sending/receiving whatever is left before closing the socket. As I mentioned before, the handles count is decremented with `close()` so that the kernel can close the socket when no handles are left.

So as long as we don't call `close()` while another thread is still sending/receiving data it shouldn't make a difference.

@ColBFritz keep in mind that calling `shutdown()` will fail with ENOTCONN if the client has already closed the connection. In that case we should catch the exception and call `close()`.

Reply:
I found that FIN vs RST thing (second answer [here](https://stackoverflow.com/questions/4160347/close-vs-shutdown-socket)). Good point about the ENOTCONN issue.

If no one objects and it doesn't cause any problems, it seems like switching back to close for now makes sense. If that causes errors, we can reassess.

Maybe it would make sense to switch everything to shutdown then close, but this would be easiest as a separate change.

Reply:
@jtpereyda thanks for the feedback, I will give it another try without shutdown() in a few days. Are "the recent sockets related fixes" you refer to #342 ? I believe I had been getting EADDRINUSE as that pull request handles so it will probably work. However, I see that it adds a minimum 5 second delay between test cases if that error is raised, whereas with shutdown it can proceed without any delay. That could greatly increase the amount of time taken to fuzz, given thousands of mutations; although in this case, it only is an issue when the target is already being restarted which _should_ be rare, and can take several seconds anyway. Still, it seems strange to just expect an error and catch it to wait for the system to clean up the socket, rather than avoiding it by explicitly shutting down the socket.

@SR4ven good catch for ENOTCONN. If the performance is acceptable without calling shutdown as jtpereyda suggests, then no specific handling is needed for that, correct?

Reply:
@ColBFritz calling shutdown before closing socket did not solve the problem with EADDRINUSE for me. If your EADDRINUSE errors are solved by calling shutdown, we can call it before every close.

Reply:
@ColBFritz you're right about the error handling. If we don't use shutdown, we don't need to care about ENOTCONN.

@starek4 I found pretty good definition of your problem [here](https://knowledgebase.progress.com/articles/Article/Can-the-time-a-socket-spends-in-TIMED-WAIT-state-be-reduced).
In that case there is no other option to speed up the fuzzing, as the kernel will take care of the closed sockets and we cannot influence that process.
According to the article, it doesn't make a difference if the socket was shutdown or closed.

@ColBFritz in the initial comment on this PR you were talking about an uncaught EAGAIN exception, while #342 fixed EADDRINUSE. Please correct me if I'm wrong but I don't think those are related. So I recommend testing this PR on top of the current master without shutdown. Please report back with the results.

We can still add shutdown to our socket closing function if needed, but we should do that consistently and in a separate PR.

Reply:
Thanks, and sorry for the delay getting back.

Rebasing onto the latest master, and removing shutdown(), I got no errors in over an hour of running, with several restarts triggered during that time.

If I also remove the close(), I get "socket.error: [Errno 98] Address already in use". I had assumed it would hit that retry code instead of throwing an uncaught exception, but it looks like #342 was only for client code, not server code.

For comparison, I ran my original code but with the shutdown() call removed, for over a half hour. I see several restarts, but still no socket errors. I worry that I might hit a corner case where close() isn't enough if I leave it running for several hours, because I'm pretty sure I saw this error before when I did not have a shutdown() call. However from the links in this discussion, even calling shutdown() may not be enough, and we could run into an issue with TIME_WAIT similar to what was fixed in #342, but on the server side.

Rather than calling shutdown(), I am adding code similar to #342 around the bind() call. I think it will work and it's probably necessary but I don't know how to force the situation where I can test it.

Reply:
> If I also remove the close(), I get "socket.error: [Errno 98] Address already in use". I had assumed it would hit that retry code instead of throwing an uncaught exception, but it looks like #342 was only for client code, not server code.

#342 was indeed only for client side connections. Currently there is no exception handling for opening server side connections.

> Rather than calling shutdown(), I am adding code similar to #342 around the bind() call. I think it will work and it's probably necessary but I don't know how to force the situation where I can test it.

Sounds like the best option. Catching `EADDRINUSE` as in the following example should do it I guess.
```python
try:
    bind...
except socket.error as e:
    if e.errno == errno.EADDRINUSE:
        raise exception.BoofuzzOutOfAvailableSockets()
```
To test this you could try to increase your `TIME_WAIT` value so that the kernel closes connections the less frequent. Combined with a high test case frequency you might be able to provoke the Address already on use error.

Reply:
I was unable to trigger the error handling code with my specific fuzzer I've been using to debug this, even after setting tcp_fin_timeout (mentioned in [the article you linked earlier](https://knowledgebase.progress.com/articles/Article/Can-the-time-a-socket-spends-in-TIMED-WAIT-state-be-reduced)) to several minutes, although I briefly saw my connection in TIME_WAIT state a few times in the netstat -a output.  I don't want to monkey with that setting too much, as this test runs on a shared resource, and this is a global setting. But, the restart code still seems to work fine, worst case with the new exception handler is that we catch an exception for a different problem and retry for a few minutes before informing the user of a connection error.